### DISCUSSION TITLE
Array proxy methods

### DISCUSSION THREAD

**Discussion by opticode** [pmndrs/valtio#124] [Q&A] [answered]
It seems that array "proxies" recognize the **push** method, but not other array modifier methods like **pop** and **splice**. Is this by design?

It is of course not hard to work around it by re-assigning the modified array, e.g.:
`state.myarray = snap.myarray.filter(...)`

**Comment by dai-shi** [comment, marked as answer]
Checking with https://github.com/pmndrs/valtio/pull/125, it seems to be working.
Do you have a more complex setup?

**Comment by opticode** [comment]
It specifically fails where the array is used in a computed value:
* `this.sections.length -= 1` -> works fine, e.g. after changing _section.length_ from 2 to 1:
```
Proxy {0: {…}}
[[Handler]]: Object
[[Target]]: Array(1)
0: {id: "987ca9a6-143b-40f3-8381-68c720f5bcf5", direction: "right", length: 25, depth: Array(2), thickness: 8}
length: 1
```
* `this.sections.pop()` -> causes error in computed value (the value that was popped from _sections_ is still in the proxy "target", but is _undefined_):

```
Proxy {0: {…}, 1: undefined}
[[Handler]]: Object
[[Target]]: Array(2)
0: {id: "fb01d14a-e1fa-4d02-a7cf-ef085f0f202e", direction: "right", length: 25, depth: Array(2), thickness: 8}
1: undefined
length: 2
```

**Reply by dai-shi** [reply]
Does this have something to do with `proxyWithComputed`?

**Reply by opticode** [reply]
I'm just calling `addComputed(state, ...` after creating my state proxy. I can confirm that pop works fine for everything other than my computed value.

**Reply by dai-shi** [reply]
Oh my, I was hoping even if `proxyWithComputed` has a limitation, `addComputed` to work.

**Reply by dai-shi** [reply]
Added a new test in #125, but still passing.
Would you create a small reproduction in codesandbox or in tests?

**Reply by opticode** [reply]
Sure, I can do that later tonight :)

**Reply by opticode** [reply]
Here you go: https://codesandbox.io/s/valtio-array-proxy-pop-computed-error-5dxhs?file=/src/App.js

**Reply by dai-shi** [reply]
Thanks. That helped.
I wanted to add computed in sync, but `arr.pop()` doesn't work with it.
Will make a fix in #125.
It's going to be one tick delay. It's not ideal, but can't be helped. `proxyWithComputed` might play well in this case. (It's unfortunate that we can't just provide just one util.)
